internal/server/data/migrations_test.go

```diff
@@ -1,231 +1,336 @@
-	assert.NilError(t, err)
-}
-
-func setupWithNoMigrations(t *testing.T, f func(db *gorm.DB)) gorm.Dialector {
-	dir := t.TempDir()
-	driver, err := NewSQLiteDriver(filepath.Join(dir, "sqlite3.db"))
-	assert.NilError(t, err)
-
-	db, err := newRawDB(driver)
-	assert.NilError(t, err)
-
-	f(db)
-
 	patch.ModelsSymmetricKey(t)
-	logging.PatchLogger(t, zerolog.NewTestWriter(t))
+	allMigrations := migrations()
 
-	return driver
-}
-
-func TestMigration_DropCertificateTables(t *testing.T) {
-	driver := setupWithNoMigrations(t, func(db *gorm.DB) {
-		loadSQL(t, db, "202206161733")
-	})
-
-	db, err := NewDB(driver, nil)
-	assert.NilError(t, err)
+	type testCase struct {
+		label    testCaseLabel
+		setup    func(t *testing.T, db *gorm.DB)
+		expected func(t *testing.T, db *gorm.DB)
+		cleanup  func(t *testing.T, db *gorm.DB)
+	}
 
-	assert.Assert(t, !tableExists(t, db, "trusted_certificates"))
-	assert.Assert(t, !tableExists(t, db, "root_certificates"))
-}
+	run := func(t *testing.T, index int, tc testCase, db *gorm.DB) {
+		logging.PatchLogger(t, zerolog.NewTestWriter(t))
+		mgs := allMigrations[:index+1]
 
-func tableExists(t *testing.T, db *gorm.DB, name string) bool {
-	t.Helper()
-	var count int
-	err := db.Raw("SELECT count(*) FROM sqlite_master WHERE type='table' AND name=?", name).Row().Scan(&count)
-	assert.NilError(t, err)
-	return count > 0
-}
+		if mID := mgs[len(mgs)-1].ID; mID != tc.label.Name {
+			t.Error("the list of test cases is not in the same order as the list of migrations")
+			t.Fatalf("test case %v was run with migration ID %v", tc.label.Name, mID)
+		}
 
-func TestMigration_AddKindToProvider(t *testing.T) {
-	for _, driver := range dbDrivers(t) {
-		t.Run(driver.Name(), func(t *testing.T) {
-			db, err := newRawDB(driver)
+		if index == 0 {
+			filename := fmt.Sprintf("testdata/migrations/%v-%v.sql", tc.label.Name, db.Dialector.Name())
+			raw, err := ioutil.ReadFile(filename)
 			assert.NilError(t, err)
 
-			loadSQL(t, db, "202206151027-"+driver.Name())
-
-			db, err = NewDB(driver, nil)
-			assert.NilError(t, err)
+			assert.NilError(t, db.Exec(string(raw)).Error)
+		}
 
-			var providers []models.Provider
-			err = db.Omit("client_secret").Find(&providers).Error
-			assert.NilError(t, err)
-			expected := []models.Provider{
-				{Name: "infra", Kind: models.ProviderKindInfra},
-				{Name: "okta", Kind: models.ProviderKindOkta, URL: "dev.okta.com"},
-			}
-			assert.DeepEqual(t, providers, expected, cmpProviderShallow)
-		})
-	}
-}
+		if tc.setup != nil {
+			tc.setup(t, db)
+		}
+		if tc.cleanup != nil {
+			defer tc.cleanup(t, db)
+		}
 
-// this test does an external call to example.okta.com, if it fails check your network connection
-func TestMigration_AddAuthURLAndScopesToProvider(t *testing.T) {
-	for _, driver := range dbDrivers(t) {
-		t.Run(driver.Name(), func(t *testing.T) {
-			db, err := newRawDB(driver)
-			assert.NilError(t, err)
+		opts := migrator.Options{
+			InitSchema: func(db *gorm.DB) error {
```

Comment:
What scenarios could cause an unexpected call to InitSchema?

Reply:
If the initial SQL file is not correct and doesn't include the migration IDs. It's unlikely, but we don't have anything else to do here, so I figured an error to indicate there is a bug in the test would be best.